fe/fe-core/src/main/java/com/starrocks/system/HistoricalNodeMgr.java

```diff
@@ -0,0 +1,103 @@
+import java.util.stream.Collectors;
+
+public class HistoricalNodeMgr {
+    private static final Logger LOG = LogManager.getLogger(HistoricalNodeMgr.class);
+    private ConcurrentHashMap<String, HistoricalNodeSet> whToComputeNodeIds;
+
+    public HistoricalNodeMgr() {
+        whToComputeNodeIds = new ConcurrentHashMap<>();
+    }
+
+    public void updateHistoricalBackendIds(List<Long> backendIds, long currentTime, String warehouse) {
+        HistoricalNodeSet nodeSet = whToComputeNodeIds.computeIfAbsent(warehouse, k -> new HistoricalNodeSet());
+        nodeSet.updateHistoricalBackendIds(backendIds, currentTime);
+    }
+
+    public void updateHistoricalComputeNodeIds(List<Long> computeNodeIds, long currentTime, String warehouse) {
+        HistoricalNodeSet nodeSet = whToComputeNodeIds.computeIfAbsent(warehouse, k -> new HistoricalNodeSet());
+        nodeSet.updateHistoricalComputeNodeIds(computeNodeIds, currentTime);
+    }
+
+    public ImmutableList<Long> getHistoricalBackendIds(String warehouse) {
+        HistoricalNodeSet nodeSet = whToComputeNodeIds.get(warehouse);
+        if (nodeSet == null) {
+            return ImmutableList.of();
+        }
+        return nodeSet.getHistoricalBackendIds();
+    }
+
+    public ImmutableList<Long> getHistoricalComputeNodeIds(String warehouse) {
+        HistoricalNodeSet nodeSet = whToComputeNodeIds.get(warehouse);
+        if (nodeSet == null) {
+            return ImmutableList.of();
+        }
+        return nodeSet.getHistoricalComputeNodeIds();
+    }
+
+    public long getLastUpdateTime(String warehouse) {
+        HistoricalNodeSet nodeSet = whToComputeNodeIds.get(warehouse);
+        if (nodeSet == null) {
+            return 0;
+        }
+        return nodeSet.getLastUpdateTime();
+    }
+
+    public ConcurrentHashMap<String, HistoricalNodeSet> getAllHistoricalNodeSet() {
+        return whToComputeNodeIds;
+    }
+
+    public void save(ImageWriter imageWriter) throws IOException, SRMetaBlockException {
+        WarehouseManager warehouseManager = GlobalStateMgr.getCurrentState().getWarehouseMgr();
+        Map<String, HistoricalNodeSet> serializedHistoricalNodes = whToComputeNodeIds.entrySet().stream()
+                .filter(entry -> warehouseManager.warehouseExists(entry.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        int numJson = 1 + serializedHistoricalNodes.size() * 2;
+        SRMetaBlockWriter writer = imageWriter.getBlockWriter(SRMetaBlockID.HISTORICAL_NODE_MGR, numJson);
+
+        writer.writeInt(serializedHistoricalNodes.size());
+        for (Map.Entry<String, HistoricalNodeSet> nodeSetEntry : serializedHistoricalNodes.entrySet()) {
+            writer.writeString(nodeSetEntry.getKey());
+            writer.writeJson(nodeSetEntry.getValue());
+        }
+        writer.close();
+        LOG.info("save image for historical node manager, serializedHistoricalNodes: {}", serializedHistoricalNodes);
```

Comment:
LOG for serializedHistoricalNodes is too huge, how about just print ids?

Reply:
> LOG for serializedHistoricalNodes is too huge, how about just print ids?

OK. Print ids is not brief and clear because they may belong to different warehouses, so I change it to only print node set size.